cpp/src/io/json/process_tokens.cu

```diff
@@ -0,0 +1,362 @@
+              } else if ('.' == c) {
+                num_state = number_state::saw_radix;
+              } else if ('e' == c || 'E' == c) {
+                num_state = number_state::start_exponent;
+              } else {
+                return false;
+              }
+              break;
+            case number_state::saw_radix:
+              if (c >= '0' && c <= '9') {
+                num_state = number_state::fraction;
+              } else if ('e' == c || 'E' == c) {
+                num_state = number_state::start_exponent;
+              } else {
+                return false;
+              }
+              break;
+            case number_state::fraction:
+              if (c >= '0' && c <= '9') {
+                num_state = number_state::fraction;
+              } else if ('e' == c || 'E' == c) {
+                num_state = number_state::start_exponent;
+              } else {
+                return false;
+              }
+              break;
+            case number_state::start_exponent:
+              if ('+' == c || '-' == c) {
+                num_state = number_state::after_sign_exponent;
+              } else if (c >= '0' && c <= '9') {
+                num_state = number_state::exponent;
+              } else {
+                return false;
+              }
+              break;
+            case number_state::after_sign_exponent:
+              if (c >= '0' && c <= '9') {
+                num_state = number_state::exponent;
+              } else {
+                return false;
+              }
+              break;
+            case number_state::exponent:
+              if (c >= '0' && c <= '9') {
+                num_state = number_state::exponent;
+              } else {
+                return false;
+              }
+              break;
+          }
+        }
+        return num_state != number_state::after_sign_exponent &&
+               num_state != number_state::start_exponent && num_state != number_state::saw_neg &&
+               num_state != number_state::saw_radix;
+      } else {
+        return false;
+      }
+    };
+
+    auto validate_strings =
+      [data = d_input.data(),
+       allow_unquoted_control_chars =
+         options.is_allowed_unquoted_control_chars()] __device__(SymbolOffsetT start,
+                                                                 SymbolOffsetT end) -> bool {
+      // This validates a quoted string. A string must match https://www.json.org/json-en.html
+      // but we already know that it has a starting and ending " and all white space has been
+      // stripped out. Also the base CUDF validation makes sure escaped chars are correct
+      // so we only need to worry about unquoted control chars
+
+      auto state   = string_state::normal;
+      auto u_count = 0;
+      for (SymbolOffsetT idx = start + 1; idx < end; idx++) {
+        auto c = data[idx];
+        if (!allow_unquoted_control_chars && c >= '\0' && c < '\32') { return false; }
```

Comment:
I saw a test failure, and I am not 100% sure why I didn't see this before. But I was able to fix it by changeing this line to.

```
        if (!allow_unquoted_control_chars && c >= 0 && c < 32) { return false; }
```

They should be identical but I think it is a difference with the compiler or options or something odd.

Reply:
No, `'\0' = 0` but `'\32' = 26'.

Reply:
I had a compiler error hence I changed it.  I will revert it back.

Reply:
```suggestion
        if (!allow_unquoted_control_chars && c >= 0 && c < 32) { return false; }
```

Reply:
Thank you. I thought typically for octal we need to prepend with `0`.  TIL about these escape sequences.